PyInstaller/utils/hooks/conda.py

```diff
@@ -181,80 +325,101 @@ def collect_dynamic_libs(name,
     """
-    file = Path(file)
+    # TODO: In theory, a package that doesn't have an ``__init__.py`` would be
+    #       missed. I've never seen a published package which didn't have an
+    #       ``__init__.py`` but it's not mandatory so it could happen.
+
+    # Get top-level folders by finding parents of `__init__.py`s
     if file.name == "__init__.py":
         file = file.parent
 
-    elif not (file.suffix in compat.ALL_SUFFIXES or file.is_dir()):
+    elif file.suffix not in compat.ALL_SUFFIXES:
+        # Skip DLLs, data files and umpteen junk files. Keep single-file
+        # packages.
         return
 
+    # Check if this file/folder's parent is in ``sys.path`` i.e. it's directly
+    # importable. This intentionally excludes submodules which would cause
+    # confusion because ``sys.prefix`` is in ``sys.path``, meaning that
+    # every file in an Conda installation is a submodule.
     for prefix in PYTHONPATH_PREFIXES:
         if len(file.parts) != len(prefix.parts) + 1:
+            # This check is redundant but speeds it up quite a bit.
             continue
+        # There are no wildcards involved here. The use of ``fnmatch`` is
+        # simply to handle the `if case-insensitive file system: use
+        # case-insensitive string matching.`
         if fnmatch.fnmatch(str(file.parent), str(prefix)):
             return file.parts[len(prefix.parts)].split(".")[0]
 
 
-def _make_package_to_distribution_map():
-    distributions_by_package_name = {}
+# All the information we want is organised the wrong way.
+
+# We want to look up distribution based on package names but we can only search
+# for packages using distribution names. And we'd like to search for a
+# distribution's json file but, due to the noisy filenames of the jsons, we can
+# only find a json's distribution rather than a distribution's json.
+
+# So we have to read everything, then regroup distributions in the ways we want
+# them grouped. This will likely be a spectacular bottleneck on full blown
+# Conda (non miniconda) with 250+ packages by default at several GiBs. I
+# suppose we could cache this on a per-json basis if it gets too much.
+
+
+def _init_distributions():
+    distributions = {}
 
     for path in CONDA_META_DIR.glob("*.json"):
-        # This will likely be quite a bottleneck on the full blown Conda (not
-        # miniconda) with 100s of packages at several GiBs. I suppose we could
-        # cache this on a per-json basis if people moan.
-        distribution = json.loads(path.read_text())
-        for file in distribution["files"]:
-            pkg = _get_package_name(file)
-            if pkg is not None:
-                distributions_by_package_name[pkg] = distribution
+        dist = Distribution(path)
+        distributions[dist.name] = dist
 
-    return distributions_by_package_name
+    return distributions
 
 
-distributions_by_package_name = _make_package_to_distribution_map()
+distributions = _init_distributions()
```

Comment:
Please use the new name in the prio commit already, to reduce diffs. Same rational as for non-functional changes.